### DISCUSSION TITLE
Please provide a sample implementation for KMM

### DISCUSSION THREAD

**Discussion by vbanavath** [joffrey-bion/krossbow#564] [General]
Hi @joffrey-bion 
I am trying to implement the Stomp client web socket utilising krossbow in KMM so that it can work on both iOS and Android
I have gone through the documentation but couldn't get much details like there is no sample implementation or some thing that i can refere for the sample implementation 

i have implemented the following thing in the KMM and trying to access it from iOSMainApp 

```kotlin
private val httpClient = HttpClient {
        install(WebSockets)
    }
    private val wsClient = KtorWebSocketClient(httpClient)
    private val stompClient = StompClient(wsClient)

    suspend fun connect() {
        val stompSession = stompClient.connect(url = baseUrl, customStompConnectHeaders = headers)
        stompSession.sendText(destination = destination, body = tripDetailsRequest.toString())
        val subscription = stompSession.subscribeText(destination = subscribe)
        subscription.collect {
            println("Received message: $it")
        }
    }
```
    
but am not able to establish the connection and getting following error 

Uncaught Kotlin exception: org.hildan.krossbow.websocket.WebSocketConnectionException: Couldn't connect to web socket at https://abcd.com/websocketApp/websocket (io.ktor.client.call.NoTransformationFoundException: Expected response body of the type 'class io.ktor.client.plugins.websocket.DefaultClientWebSocketSession' but was 'class io.ktor.utils.io.ByteChannelNative'
In response from `https://abcd.com/websocketApp/websocket`
Response status `400 Bad Request`
Response header `ContentType: null`
Request header `Accept: */*`

i have added the dependencies as followed in commonMain
```
implementation("org.hildan.krossbow:krossbow-websocket-ktor:7.0.0")
implementation("org.hildan.krossbow:krossbow-stomp-kxserialization-json:7.0.0")
implementation("io.ktor:ktor-client-cio:2.3.10")
implementation("io.ktor:ktor-client-websockets:2.3.12")
```

quick and use full response is very much appreciated!
thanks

**Comment by joffrey-bion** [comment]
Hi! Thanks for reaching out.

Given that you're getting an HTTP 400, it should be a problem in your initial connection request. There is a chance this is due to the `https://` scheme. Have you tried using `ws://` or `wss://` for your URL?

**Reply by vbanavath** [reply]
Hi! Thanks for the quick response 
but our server supports only https endpoint we don't have a support for `ws or wss` 
Interestingly we tried same thing with iOS native sample(https://github.com/Romixery/SwiftStomp) and its working there we passed as the `https` only
when we try to achieve the same thing in KMM its throwing the above error

**Reply by vbanavath** [reply]
i have observed that the session here is not considering the passing headers for the initial communication 

```
suspend fun connect(
        url: String,
        login: String? = null,
        passcode: String? = null,
        host: String? = DefaultHost,
        customStompConnectHeaders: Map<String, String> = emptyMap(),
        sessionCoroutineContext: CoroutineContext = EmptyCoroutineContext,
    ): StompSession {
        val session = withTimeoutOrNull(config.connectionTimeout) {
            **val webSocket = webSocketClient.connect(
                url = url,
                protocols = StompVersion.preferredOrder.map { it.wsSubprotocolId },
            )**
            webSocket.stomp(
                config = config,
                host = host,
                login = login,
                passcode = passcode,
                customHeaders = customStompConnectHeaders,
                sessionCoroutineContext = sessionCoroutineContext,
            )
        }
        return session ?: throw ConnectionTimeout(url, config.connectionTimeout)
    }
```

webSocketClient is only taking the url and protocols not the passing headers
can we overrode this `webSocketClient.connect`

**Reply by joffrey-bion** [reply]
`customStompConnectHeaders` are headers at the stomp protocol level (they are sent in the STOMP frame, not in the HTTP handshake). If you want to customize the handshake, you can make the handshake separately first, and then use the `.stomp()` extension function. See the documentation here:
https://joffrey-bion.github.io/krossbow/stomp/advanced-features/#using-custom-headers-in-the-web-socket-handshake

**Reply by vbanavath** [reply]
we have included map of following headers `"Upgrade" to "websocket","Connection" to "upgrade", along with required one in the but the server says the following error 
2024-09-05 11:21:27.370 ERROR [CONNECTED_ORCHESTRATION,dfb858ec6edc18f1,dfb858ec6edc18f1,true] 1 --- [o-1414-exec-222] o.s.w.s.s.s.DefaultHandshakeHandler      : Handshake failed due to invalid Upgrade header: null

failed with exception: io.ktor.http.UnsafeHeaderException: Header(s) [Upgrade] are controlled by the engine and cannot be set explicitly

**Reply by joffrey-bion** [reply]
Yes, you should only set your _additional_ custom headers for the handshake, not the standard websocket handshake headers, as these are taken care of by the websocket implementations. What custom headers did you try to send? Does your server expect some extra headers? If not, there should be no need to customize that.

Another option for getting a 400 could be that your server is expecting the SockJS protocol, but you're connecting using regular websocket protocol. Could you check that? If your spring server has SockJS enabled, you might want to remove that.
